core/src/main/java/io/kestra/core/server/ServiceLivenessManager.java

```diff
@@ -0,0 +1,380 @@
+                                                 final ServiceInstance instance) {
+        return true; // noop
+    }
+
+    /**
+     * Gets the Instant of the last time the state was successfully updated.
+     *
+     * @return the {@link Instant}.
+     */
+    protected Instant lastSucceedStateUpdated() {
+        return lastSucceedStateUpdated;
+    }
+
+
+    /**
+     * Updates the state of the local worker instance.
+     *
+     * @param newState           the new worker state.
+     * @param onStateChangeError the callback to invoke if the state cannot be changed.
+     */
+    protected ServiceInstance updateServiceInstanceState(final Instant now,
+                                                         final Service service,
+                                                         final Service.ServiceState newState,
+                                                         final OnStateTransitionFailureCallback onStateChangeError) {
+
+        // Pre-check the state transition validation with the known local state.
+        ServiceInstance localInstance = localService(service).instance();
+        if (!localInstance.state().isValidTransition(newState)) {
+            log.warn("Failed to transition service [id={}, type={}, hostname={}] from {} to {}. Cause: {}.",
+                localInstance.id(),
+                localInstance.type(),
+                localInstance.server().hostname(),
+                localInstance.state(),
+                newState,
+                "Invalid transition"
+            );
+            mayDisableStateUpdate(service, localInstance);
+            return localInstance;
+        }
+
+        // Ensure only one thread can update any instance at a time.
+        stateLock.lock();
+        // Optional callback to be executed at the end.
+        Runnable returnCallback = null;
+        try {
+            localInstance = localService(service).instance();
+            ServiceStateTransition.Response response = serviceRepository.mayTransitionServiceTo(localInstance, newState);
+
+            ServiceInstance remoteInstance = response.instance();
+
+            boolean isStateTransitionSucceed = response.is(Result.SUCCEED);
+
+            if (response.is(Result.ABORTED)) {
+                // Force state transition due to inconsistent state; remote state does not exist (yet).
+                remoteInstance = serviceRepository.save(localInstance.updateState(newState, now));
+                isStateTransitionSucceed = true;
+            }
+
+            if (response.is(Result.FAILED)) {
+                if (remoteInstance.seqId() < localInstance.seqId()) {
+                    // Force state transition due to inconsistent state; remote state is not up-to-date.
+                    remoteInstance = serviceRepository.save(localInstance.updateState(newState, now));
+                    isStateTransitionSucceed = true;
+                } else {
```

Comment:
I didn't dig into the code but it seems to be a read then write pattern?
If yes, you should SELECT FOR UPDATE and use a transaction (via a transactional callback) or there may be concurrency issue

Reply:
You right;  the `mayTransitionServiceTo` do a read/write pattern (the JDBC implementation used a SELECT for update) . Moreover, this code is used independently of the repository implementation so transactions are not always available. Also, as we have mostly a single-writer pattern race conditions should not happen so it seems to be acceptable (most of the time only the service is update its own state).